RecoLocalTracker/SiPixelRecHits/src/PixelCPEFast.cc

```diff
@@ -0,0 +1,594 @@
+    g.frame = pixelCPEforGPU::Frame(vv.x(), vv.y(), vv.z(), rr);
+
+    zl += vv.z();
+    miz = std::min(miz, std::abs(vv.z()));
+    mxz = std::max(mxz, std::abs(vv.z()));
+    rl += vv.perp();
+    pl += vv.phi();  // (not obvious)
+
+    // errors .....
+    ClusterParamGeneric cp;
+    auto gvx = p.theOrigin.x() + 40.f * m_commonParamsGPU.thePitchX;
+    auto gvy = p.theOrigin.y();
+    auto gvz = 1.f / p.theOrigin.z();
+    //--- Note that the normalization is not required as only the ratio used
+
+    // calculate angles
+    cp.cotalpha = gvx * gvz;
+    cp.cotbeta = gvy * gvz;
+
+    cp.with_track_angle = false;
+
+    auto lape = p.theDet->localAlignmentError();
+    if (lape.invalid())
+      lape = LocalError();  // zero....
+
+#ifdef DUMP_ERRORS
+    auto m = 10000.f;
+    for (float qclus = 15000; qclus < 35000; qclus += 15000) {
+      errorFromTemplates(p, cp, qclus);
+
+      std::cout << i << ' ' << qclus << ' ' << cp.pixmx << ' ' << m * cp.sigmax << ' ' << m * cp.sx1 << ' '
+                << m * cp.sx2 << ' ' << m * cp.sigmay << ' ' << m * cp.sy1 << ' ' << m * cp.sy2 << std::endl;
+    }
+    std::cout << i << ' ' << m * std::sqrt(lape.xx()) << ' ' << m * std::sqrt(lape.yy()) << std::endl;
+#endif
+
+    errorFromTemplates(p, cp, 20000.f);
+    g.pixmx = std::max(0, cp.pixmx);
+    g.sx[0] = cp.sigmax;
+    g.sx[1] = cp.sx1;
+    g.sx[2] = cp.sx2;
+
+    g.sy[0] = cp.sigmay;
+    g.sy[1] = cp.sy1;
+    g.sy[2] = cp.sy2;
+
+    /*
+    // from run1??
+    if (i<96) {
+      g.sx[0] = 0.00120;
+      g.sx[1] = 0.00115;
+      g.sx[2] = 0.0050;
+
+      g.sy[0] = 0.00210;
+      g.sy[1] = 0.00375;
+      g.sy[2] = 0.0085;
+    } else if (g.isBarrel) {
+      g.sx[0] = 0.00120;
+      g.sx[1] = 0.00115;
+      g.sx[2] = 0.0050;
+
+      g.sy[0] = 0.00210;
+      g.sy[1] = 0.00375;
+      g.sy[2] = 0.0085;
+   } else {
+      g.sx[0] = 0.0020;
+      g.sx[1] = 0.0020;
+      g.sx[2] = 0.0050;
+
+      g.sy[0] = 0.0021;
+      g.sy[1] = 0.0021;
+      g.sy[2] = 0.0085;
+   }
+   */
+
+    for (int i = 0; i < 3; ++i) {
+      g.sx[i] = std::sqrt(g.sx[i] * g.sx[i] + lape.xx());
+      g.sy[i] = std::sqrt(g.sy[i] * g.sy[i] + lape.yy());
+    }
+  }
+
+  // compute ladder baricenter (only in global z) for the barrel
+  auto& aveGeom = m_averageGeometry;
+  int il = 0;
+  for (int im = 0, nm = phase1PixelTopology::numberOfModulesInBarrel; im < nm; ++im) {
+    auto const& g = m_detParamsGPU[im];
+    il = im / 8;
+    assert(il < int(phase1PixelTopology::numberOfLaddersInBarrel));
+    auto z = g.frame.z();
+    aveGeom.ladderZ[il] += 0.125f * z;
+    aveGeom.ladderMinZ[il] = std::min(aveGeom.ladderMinZ[il], z);
+    aveGeom.ladderMaxZ[il] = std::max(aveGeom.ladderMaxZ[il], z);
+    aveGeom.ladderX[il] += 0.125f * g.frame.x();
+    aveGeom.ladderY[il] += 0.125f * g.frame.y();
+    aveGeom.ladderR[il] += 0.125 * sqrt(g.frame.x() * g.frame.x() + g.frame.y() * g.frame.y());
```

Comment:
```suggestion
    aveGeom.ladderR[il] += 0.125f * sqrt(g.frame.x() * g.frame.x() + g.frame.y() * g.frame.y());
```
?

Reply:
Done in https://github.com/cms-patatrack/cmssw/pull/593/commits/9dc393168636f3c4e48986f7739c554f75cdd1c1 .